### PR TITLE
THES-63: Create Order (from LMS) - Step 2 - Add Connections to Coordinator

### PR DESCRIPTION
**Pull request by grmartin**
Adding functionality as required by THES-63:

- THES-63: Create Order (from LMS) - Step 2 - Add Connections to Coordinator
  - THES-133: Implement OrderView in lms/views.py
    This will be almost a exact copy and paste of commerce_coordinator.apps.ecommerce.views.OrderView.
  
    **Implementation Notes**
  
  - We are changing from a backend-to-backend endpoint to a user-facing endpoint.
    - ecommerce’s OrderView was meant to be an administrative endpoint where ecommerce sent the user details for the created order to Coordinator.
    - However, lms' OrderView is different because the user is hitting the endpoint, not a backend system.
    - See TODOs in the code, but this means:
      - [x] Need to change permission_classes to edx_rest_framework_extensions.permissions.IsAuthenticated [[DRF docs](https://www.django-rest-framework.org/api-guide/permissions/#isauthenticated)]
      - [x] Need to read in the user’s information from the request, not from the parameters
      - [?] Need to convert user’s id from a Coordinator id to a lms_user_id
      - [x] We should be able to re-use ecommerce’s serializer to validate the data
      - [X] Need to redirect the user to Payment MFE.
        - [X] May want to do something similar to what ecommerce does,
  - See Connections sheet for links to current code in lms and ecommercce. 
  - THES-134: Implement OrderCreateRequested in lms/filters.py
  - THES-135: Implement CreateTitanOrder in titan/pipeline.py
  - THES-136: Implement TitanAPIClient.create_order in titan/clients.py

### CODE REVIEW

File: commerce_coordinator/apps/titan/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    def run_filter(self, params, order_data):  # pylint: disable=arguments-differ
+        """
+        Execute a filter with the signature specified.
+        Arguments:
+            params: arguments passed through from the original order history url querystring
+            order_data: any preliminary orders (from earlier pipeline step) we want to append to
+        """
+
+        titan_api_client = TitanAPIClient()
+        titan_response = titan_api_client.create_order(**params)
```
Comment (HammadAhmadWaqas) [inline commerce_coordinator/apps/titan/pipeline.py:24]: We are not calling the correct `create_order` here.

There should be a new `create_order` implement as part of the story [THES-136](https://2u-internal.atlassian.net/browse/THES-136)
Seem like that method is missing from this PR.

updated create_order method should call 3 Titan's APIs as shown in this [Mermaid](https://mermaid.live/edit#pako:eNrNVl1v2jAU_SvIz-A0kKTgh0kboKkPUyuK-jBFirz4AlYTO7MdWob63-eYpd2qAkajUvPkXPt86NyrxFuUSwaIoEUhH_IVVaYz_5IKXf9YKlqtOteKgepMwFBe6FR07MO4gtxwKdzJppLLsgSVQ5ZLqRgX1EiFaVVpDO0W1nwpaKGxbAizXAE1wLJdtdPrfTpAYrihYh8BXYO_iVpZgpUxlSZB0FZ7fwOAPVqFZfCMCZweIV-n8yM2X2TWHB40dtHd2aW_v1dAb8FD4R5TL0qNF7wwoP4ojx18Bj9r0JbGqzkVr6DgAvAOPG-KjsxH3b8r9rR_PxrqEzvxBsRD5FB676O4Q7SRVXRTgjBtSoTMppOr2XQ8Pya-611ecIvW2DXt883V2L1jyljGDZRe_W85WktLMLq0HxNQ2EID6yygFQ_WYZDbatByE3Jzfft_Ni2sKsBA5ubi3GZXkN_L2gStyjkMuwHJmhzeIdrzGTw9zz0D9CHM_DsmH8PSyyD4-dn7kT2DixNi2f8j9vJhqL5_C5w1G34ODlKcLwwQDHWR5SkpZ_aKtG3cpcisoIQUEbtksKB1YVKUiid7lNZG3m5EjsjCBgRdVFfM8k04tZep8lV1yrg19VysqEBkix4R6Q1jHMVRPw6jfn94MQjDLtogEvdxEo2iOAkHyeVwGF5GT130S0rLG-L4Ypgkg9EoDpOkH8UDx_fdbRpVW3pwat92dz135Xv6DcSGq8E)

As we are doing [here](https://github.com/edx/commerce-coordinator/blob/563d09dd4828282a3ce071f5250a21b32b8a60b4/commerce_coordinator/apps/titan/tasks.py#L36) in `order_created_save_task`.
Then `order_created_save_task` should be updated to use this new order_create.

existing `order_create` should be renamed to `create_cart` (see diagram)

File: commerce_coordinator/apps/lms/serializers.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    """
+    Serializer for order_created_signal input validation.
+    """
+    product_sku = serializers.ListField(
+        child=serializers.CharField(), allow_empty=False
+    )
+    edx_lms_user_id = serializers.IntegerField(allow_null=False)
+    email = serializers.EmailField(allow_null=False)
+    first_name = serializers.CharField(allow_null=False)
+    last_name = serializers.CharField(allow_null=False)
```
Comment (HammadAhmadWaqas) [inline commerce_coordinator/apps/lms/serializers.py:16]: These 2 fields are nullable in the User model (in LMS as well as in Coordinator). So we not make them mandatory.

We just need to confirm from Titans that their cart creation API will not through an error for not providing these fields.

File: commerce_coordinator/settings/local.py [hunk 3]
```diff
@@ -118,10 +118,14 @@
 
 EDX_API_KEY = 'PUT_YOUR_API_KEY_HERE'  # This is the actual API key in devstack.
 
+# DevStack: https://edx.readthedocs.io/projects/open-edx-devstack/en/latest/service_list.html
 ECOMMERCE_URL = "http://localhost:18130"
 
 LMS_URL_ROOT = "http://localhost:18000"
 
+PAYMENT_MICROFRONTEND_URL = "http://localhost:1998"
+# /DevStack
```
Comment (HammadAhmadWaqas) [inline commerce_coordinator/settings/local.py:127]: same here

**Review by HammadAhmadWaqas — CHANGES_REQUESTED** [review]

File: commerce_coordinator/apps/lms/views.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            'first_name': request.user.first_name if IS_LIVE else 'John',
+            'last_name': request.user.last_name if IS_LIVE else 'Doe',
+            'coupon_code': request.query_params.get('coupon_code'),
+        }
+        serializer = OrderCreatedSignalInputSerializer(data=order_created_signal_params)
+
+        if serializer.is_valid(raise_exception=True):
+            result = OrderCreateRequested.run_filter(serializer.validated_data)
+            logger.debug(f'{self.get.__qualname__} pipeline result: {result}.')
+            return self._redirect_response_to_basket_or_payment(request)
```
[the remark this answers is not in the document]
Comment (grmartin) [reply commerce_coordinator/apps/lms/views.py:65]: Is there a way to know if the result is in error? For example, will it return an Exception-like object?

File: commerce_coordinator/settings/local.py [hunk 5]
```diff
@@ -118,10 +118,14 @@
 
 EDX_API_KEY = 'PUT_YOUR_API_KEY_HERE'  # This is the actual API key in devstack.
 
+# DevStack: https://edx.readthedocs.io/projects/open-edx-devstack/en/latest/service_list.html
```
[the remark this answers is not in the document]
Comment (grmartin) [reply commerce_coordinator/settings/local.py:121]: just a note that these are URLs to other DevStack services and the complete list is here.

File: commerce_coordinator/apps/core/tests/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    @ddt.data(
+        {"x": [1, 2, 3], "t": list, "class_name": "WrappedList"},
+        {"x": ["1", 2, 3], "t": list, "class_name": "WrappedList"},
+        {"x": ["1", {}, (), []], "t": list, "class_name": "WrappedList"},
+        {"x": tuple((1, 2, 3)), "t": tuple, "class_name": "WrappedTuple"},
+        {"x": tuple(("1", 2, 3)), "t": tuple, "class_name": "WrappedTuple"},
+        {"x": tuple(("1", {}, (), [])), "t": tuple, "class_name": "WrappedTuple"},
+        {"x": {"a": 1, "b": 2, "c": 3}, "t": dict, "class_name": "WrappedDict"},
+        {"x": {"a": "1", "b": 2, "c": 3}, "t": dict, "class_name": "WrappedDict"},
+        {"x": {"a": "1", "b": {}, "c": (), "d": []}, "t": dict, "class_name": "WrappedDict"},
```
[the remark this answers is not in the document]
Comment (grmartin) [reply commerce_coordinator/apps/core/tests/test_utils.py:23]: a b c d are just to prove types, but yeah ill break out the rest so its clearer thats test data.

File: commerce_coordinator/apps/lms/views.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        """
+        logger.debug(f'{self.get.__qualname__} request object: {request.data}.')
+        logger.debug(f'{self.get.__qualname__} headers: {request.headers}.')
+
+        order_created_signal_params = {
+            'product_sku': request.query_params.getlist('product_sku'),
+            'edx_lms_user_id': request.user.lms_user_id if IS_LIVE else request.query_params.get('edx_lms_user_id'),
+            'email': request.user.email if IS_LIVE else request.query_params.get('email'),
+            'first_name': request.user.first_name if IS_LIVE else 'John',
+            'last_name': request.user.last_name if IS_LIVE else 'Doe',
```
[the remark this answers is not in the document]
Comment (grmartin) [reply commerce_coordinator/apps/lms/views.py:57]: ill have to update the tests to fully fill in the user.

File: commerce_coordinator/apps/lms/serializers.py (hunk 2 above)
Comment (grmartin) [reply commerce_coordinator/apps/lms/serializers.py:16]: ive removed the allow null prohibition for now while i look in to the down line on this.

File: commerce_coordinator/apps/lms/views.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (grmartin) [reply commerce_coordinator/apps/lms/views.py]: Is there something similar to this?

```python

            if isinstance(result, Exception):
                # DO X
                return return_exception_as_message(result)
            elif is_error(result):
                # DO Y
                return return_error(result)
            else:
                return self._redirect_response_payment(request)
```

[the remark this answers is not in the document]
Comment (HammadAhmadWaqas) [reply commerce_coordinator/apps/lms/views.py]: Yes, For Exceptions we have to use the `try` and `except` blocks, Example: [BasketAddItemsView](https://github.com/openedx/ecommerce/blob/7555353ae972563fd293558eea608ae6151c1186/ecommerce/extensions/basket/views.py#L437)

File: commerce_coordinator/apps/lms/views.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (grmartin) [reply commerce_coordinator/apps/lms/views.py]: This is now pulling from the user object and tests are updated to comply.

File: commerce_coordinator/apps/lms/views.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (grmartin) [reply commerce_coordinator/apps/lms/views.py]: Ok, i now have this tested and working (i made an exploding pipeline) see https://github.com/edx/commerce-coordinator/pull/73/files#diff-d887746bdaeac8ad9988169c12b3c8c6f59754d7c6977c664980a1d53a58af97R22

File: commerce_coordinator/apps/core/tests/test_utils.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        (500, True),
+        (1000, True)
+    )
+    @ddt.unpack
+    def test_random_unicode_str_randomness_over_time(self, strlen, limit_unicode):
+        tries = 1000
+
+        examples = dict(
+            (h, True)  # kvp, val (True) is ignored
+            for h in [
```
Comment (HammadAhmadWaqas) [inline commerce_coordinator/apps/core/tests/test_utils.py:58]: Descriptive names are encouraged in Python. [reference](https://realpython.com/python-pep8/#how-to-choose-names).

File: commerce_coordinator/apps/core/tests/utils.py [hunk 9]
```diff
@@ -232,3 +233,68 @@ def assertJSONClientResponse(
...
+    class WrappedTuple(tuple):
+        pass
+
+    class WrappedList(list):
+        pass
+
+    class WrappedDict(dict):
+        pass
+
+    wx = None
```
Comment (HammadAhmadWaqas) [inline commerce_coordinator/apps/core/tests/utils.py:255]: what does this `wx` stands for?

File: commerce_coordinator/apps/lms/views.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+        Args:
+            request (django.http.HttpRequest): django.http.HttpRequest
+
+        Query Parameters:
+            product_sku: Array. An edx.org stock keeping units (SKUs) that the user would like to purchase.
+            coupon_code: (Optional) A coupon code to initially apply to the order.
+            edx_lms_user_id: (Temporary) Initially we will be calling this API from a server. this param is to bypass
+                the edx_lms_user_id from the calling server. later on, we will remove this param and extract
+                edx_lms_user_id from request.user.lms_user_id.
```
[the remark this answers is not in the document]
Comment (grmartin) [reply commerce_coordinator/apps/lms/views.py:38]: Really?

File: commerce_coordinator/apps/ecommerce/tests/test_views.py [hunk 11]
```diff
@@ -57,54 +59,46 @@ def test_view_rejects_unauthorized(self):
         self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)
 
     @ddt.data(
-        (
-            # test success.
+        name_test("test success", (
```
[the remark this answers is not in the document]
Comment (grmartin) [reply commerce_coordinator/apps/ecommerce/tests/test_views.py:62]: im trying. :)

File: commerce_coordinator/apps/core/tests/utils.py (hunk 9 above)
Comment (grmartin) [reply commerce_coordinator/apps/core/tests/utils.py:255]: WrappedX ofc. ;)

File: commerce_coordinator/apps/core/tests/utils.py [hunk 12]
```diff
@@ -232,3 +233,68 @@ def assertJSONClientResponse(
             mock_status=mock_status,
             expected_output=expected_output
         )
+
+
+def name_test(name: str, x):
```
[the remark this answers is not in the document]
Comment (grmartin) [reply commerce_coordinator/apps/core/tests/utils.py:238]: ill make this clearer, sorry, coming from C# where `x` is assumed to be `subject`.

File: commerce_coordinator/apps/core/tests/test_utils.py (hunk 8 above)
Comment (grmartin) [reply commerce_coordinator/apps/core/tests/test_utils.py:58]: ill find a way to be more clear

File: commerce_coordinator/apps/lms/views.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (grmartin) [reply commerce_coordinator/apps/lms/views.py:38]: This API reads it from the user object.

File: commerce_coordinator/apps/titan/pipeline.py (hunk 1 above)
Comment (grmartin) [reply commerce_coordinator/apps/titan/pipeline.py:24]: I think (according to tests) performs as expected. Im working on a final validation in the next day.

File: commerce_coordinator/apps/lms/serializers.py (hunk 2 above)
Comment (grmartin) [reply commerce_coordinator/apps/lms/serializers.py]: We may need to fix this forward if they say no (as in currently).

Comment (grmartin) [reply commerce_coordinator/apps/lms/serializers.py]: Can we catchup tomorrow on exactly what you mean? (incase im understanding wrongly)

Comment (grmartin) [reply commerce_coordinator/apps/lms/serializers.py]: Question asked, https://twou.slack.com/archives/C03S3D8736H/p1684430310470109

(sorry pasted a message i had prepped for another field, so you get 2 emails)

**Comment by grmartin** [comment]
Note to self: I already updated rollout docs (as this is a reused identifier) i dont expect it to change.... if it does i need to fix it and if its approved i need to remove the warning comment.

**Review by HammadAhmadWaqas — APPROVED** [review]
LGTM 👍

**Pull request merged**